Sap odp

sumitnigamcs opened the pull request:
Please review

sahilkapoor301992 committed:
ODP Changes

ksahilCF committed:
Fix Code

ksahilCF committed:
Jar Password commented

ksahilCF committed:
updated RFM changes

sumitnigamcs committed:
refactoring

sumitnigamcs committed:
temp file removed

sumitnigamcs committed:
checkstyle

sumitnigamcs committed:
review comments adjusted

itsankit-google commented on src/e2e-test/java/io/cdap/plugin/odp/locators/ODPLocators.java:
```diff
@@ -0,0 +1,148 @@
+ */
+public class ODPLocators {
+
+  @FindBy(how = How.XPATH, using = "//*[contains(text(),'SAP ODP')]")
```
should use something like `@FindBy(how = How.XPATH, using = "//*[@data-cy=\"plugin-SapOdp-batchsource\"]")` here

sumitnigamcs replied:
done

itsankit-google commented on src/e2e-test/java/io/cdap/plugin/odp/locators/ODPLocators.java:
```diff
@@ -0,0 +1,148 @@
+  @FindBy(how = How.XPATH, using = "//*[contains(text(),'SAP ODP')]")
+  public static WebElement sapODPSource;
+
+  @FindBy(how = How.XPATH, using = "//*[contains(text(),'-SNAPSHOT')]")
```
can this locator be made more specific so that it is used to locate properties of SAP ODP plugin specifically?because there can be other plugins as well which can match this.

sumitnigamcs replied:
done

itsankit-google commented on src/e2e-test/java/io/cdap/plugin/odp/stepsdesign/DesginTimeODP.java:
```diff
@@ -0,0 +1,135 @@
+  @When("User has selected Sap client macro to configure")
+  public void userHasSelectedSapClientMacroToConfigure() {
+    ODPLocators.macros.get(0).click();
+    ODPLocators.sapClient.sendKeys(Keys.chord(Keys.CONTROL, A));
```
concerned if this operation is OS specific because in MAC we use `CMD+A` to perform this operation?
If the use case is to clear the input text and add value can we use `SeleniumHelper.replaceElementValue()` for this? @rmstar what do you think?

sumitnigamcs replied:
done

itsankit-google commented on src/e2e-test/java/io/cdap/plugin/odp/stepsdesign/DesginTimeODP.java:
```diff
@@ -0,0 +1,135 @@
+  @When("User has selected Sap language macro to configure")
+  public void userHasSelectedSapLagMacroToConfigure() {
+    ODPLocators.macros.get(1).click();
+    ODPLocators.language.sendKeys(Keys.chord(Keys.CONTROL, A));
```
similar comment as above in the whole PR wherever `(Keys.CONTROL, A)` is used.

sumitnigamcs replied:
done

itsankit-google commented on src/e2e-test/java/io/cdap/plugin/odp/locators/ODPLocators.java:
```diff
@@ -0,0 +1,148 @@
+  public static WebElement validateButton;
+
+  @FindBy(how = How.XPATH, using = "//*[@data-cy='plugin-validation-success-msg']")
+  public static WebElement successMessage;
```
successMessage -> pluginValidationSuccessMsg

sumitnigamcs replied:
done

itsankit-google commented on src/e2e-test/java/io/cdap/plugin/odp/stepsdesign/DesginTimeODP.java:
```diff
@@ -0,0 +1,135 @@
+
+  @When("User has selected Sap server as host macro to configure")
+  public void userHasSelectedSapServerAsHostMacroToConfigure() {
+    ODPLocators.macros.get(2).click();
```
Is there any better way to handle this? This makes the test brittle. In future, if the plugin configs order is changed then whole test will break! @rmstar What do you think?

ksahilCF replied:
@itsankit-google I dont think so we can handle it any other way as these macros are in a single class only the index is the difference. If order changes then we need to again change the order of identification

rmstar replied:
yeah seems brittle.. what happens if a new config param is added? would that affect the index of other macros?
would be good if we can find a better way to handle it.

ksahilCF replied:
got the solution for every macro element we can capture an object path will remain constant- //*[@data-cy="jco.client.ashost"]//child::button//span//span//strong only the data-cy value will be changed. @sumitnigamcs please capture all the macros seprately and call via name

sumitnigamcs replied:
done

itsankit-google commented on src/e2e-test/java/io/cdap/plugin/odp/stepsdesign/DesginTimeODP.java:
```diff
@@ -0,0 +1,135 @@
+
+  @When("User has selected System Number macro to configure")
+  public void userHasSelectedSystemNumberMacroToConfigure() {
+    ODPLocators.macros.get(3).click();
```
similar comment as above in the whole PR wherever macros are used.

sumitnigamcs replied:
done

itsankit-google commented on src/e2e-test/java/io/cdap/plugin/odp/stepsdesign/RuntimeODP.java:
```diff
@@ -0,0 +1,396 @@
+    InputStream input;
+    Properties connection = new SAPProperties();
+    input = new FileInputStream("src/e2e-test/resources/Google_SAP_Connection.properties");
+    prop.load(input);
```
can we load property file once instead on every call?

sumitnigamcs replied:
done

itsankit-google commented on src/e2e-test/java/io/cdap/plugin/odp/stepsdesign/RuntimeODP.java:
```diff
@@ -0,0 +1,396 @@
+  @Then("Get Count of no of records transferred from ODP to BigQuery in {string}")
+  public void getCountOfNoOfRecordsTransferredFromODPToBigQueryIn(String table) throws
+    IOException, InterruptedException {
+    countRecords = gcpClient.countBqQuery(CDAPUtils.getPluginProp(table));
```
Here, we should also verify the metric in the UI, no of records out from the source to the no of records in sink.

sumitnigamcs replied:
will fix this in CDAP-18634

ksahilCF committed:
latest changes RFM

sumitnigamcs committed:
review comments

sumitnigamcs committed:
merge conflicts

sumitnigamcs committed:
locators modified

rmstar commented on pom.xml:
```diff
@@ -138,6 +139,11 @@
       <scope>provided</scope>
     </dependency>
+    <dependency>
+      <groupId>com.google.guava</groupId>
```
why are we adding a guava dependency here?

ksahilCF replied:
I tried adding inside e2e-tests profile but it was not getting resolved there.

tivv replied:
What are the issues? The way it's not it's not going to fly. Guava versioning is pretty tricky, so we must not add it to production code unless needed by production code

sumitnigamcs replied:
Hi @tivv I have removed guava from main dependency section, moved it to profile's dependency management, along with adding it to classpath via profile's maven-dependency-plugin. Please confirm if this is fine

tivv commented on pom.xml:
```diff
@@ -510,6 +521,137 @@
+            <version>3.0.0-M5</version>
+            <configuration>
+              <includes>
+                <include>TestRunner.java, TestRunnerMand.java, TestRunnerOpt.java</include>
```
Use pattern instead of listing all classes

sumitnigamcs replied:
@tivv since we have to maintain the execution order of the TestRunners, we had to add the runner filenames like this. We have to make sure TestRunnerOpt runs in the end, else it will impact TestRunnerMand

tivv replied:
Why do we need an execution order? It sounds like an issue. All tests must be independent. One should be able to run any test separately or tests in any sequence. That's basic principle.

sumitnigamcs replied:
@tivv For config driven scenarios, wherein we are implementing cucumberJvm plugin, we are loading optional properties on the run, once they are loaded, the scenarios with only mandatory fields will not run. So mandatory field scenario should run first, then plugin loading the optional properties should execute followed by optional field scenarios. Suggest if I am not able to put forward my point, or if I am missing something to make the scenarios more independent

tivv replied:
Sure, could you please clarify in more details. Overall approach is that tests must be self-contained and do any necessary clean-up as needed. Why is it not possible here?

sumitnigamcs replied:
done, thanks

rmstar commented on pom.xml:
```diff
@@ -510,6 +521,137 @@
         </executions>
       </plugin>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
```
why are these added outside e2e-tests profile?

sumitnigamcs replied:
removed, these were rather duplicate

rmstar commented on src/e2e-test/java/io/cdap/plugin/odp/stepsdesign/DesginTimeODP.java:
```diff
@@ -0,0 +1,121 @@
+/**
+ * DesginTimeODP.
+ */
+public class DesginTimeODP {
```
nit: spelling

sumitnigamcs replied:
done

rmstar commented on src/e2e-test/java/io/cdap/plugin/odp/stepsdesign/DesginTimeODP.java:
```diff
@@ -0,0 +1,121 @@
+  @When("User has selected Sap msServ macro to configure")
+  public void userHasSelectedSapMsServMacroToConfigure() {
+    ODPLocators.macros.get(4).click();
+    SeleniumHelper.replaceElementValue(ODPLocators.portNumber, "${portN0}");
```
Why is it N0 (zero instead of 'O')

sumitnigamcs replied:
done

rmstar commented on src/e2e-test/java/io/cdap/plugin/odp/stepsdesign/DesginTimeODP.java:
```diff
@@ -0,0 +1,121 @@
+  @When("User has selected Sap msHost macro to configure")
+  public void userHasSelectedSapMsHostMacroToConfigure() {
+    ODPLocators.macros.get(2).click();
+    SeleniumHelper.replaceElementValue(ODPLocators.msHost, "${msHOst}");
```
${msHost} ?

ksahilCF replied:
Thats a macro variable which we have taken.

sumitnigamcs replied:
done

rmstar commented on src/e2e-test/java/io/cdap/plugin/odp/stepsdesign/DesignTimeODPError.java:
```diff
@@ -0,0 +1,118 @@
+
+  @Then("User is able to validate the validate the error")
+  public void userIsAbleToValidateTheValidateTheError() {
+    Assert.assertTrue(errorExist);
```
this assertion can be moved into userIsAbleToSetParameterAsAndGetting().

sumitnigamcs replied:
done

rmstar commented on src/e2e-test/java/io/cdap/plugin/odp/stepsdesign/DesignTimeODPError.java:
```diff
@@ -0,0 +1,118 @@
+  static String color;
+
+  @Then("{string} as {string} and getting {string}")
+  public void userIsAbleToSetParameterAsAndGetting(String arg0, String input, String errorMessage) {
```
please rename. userIsAbleToSetParameterAsAndGetting doesn't make sense

sumitnigamcs replied:
done

rmstar commented on src/e2e-test/java/io/cdap/plugin/odp/stepsdesign/DesignTimeODPError.java:
```diff
@@ -0,0 +1,118 @@
+  }
+
+  @Then("^User is able to set parameter (.+) as (.+) and getting row (.+) for wrong input$")
+  public void user_is_able_to_set_parameter_as_and_getting_row_for_wrong_input(
```
use camelcase

sumitnigamcs replied:
done

rmstar commented on src/e2e-test/java/io/cdap/plugin/odp/stepsdesign/RuntimeODP.java:
```diff
@@ -0,0 +1,422 @@
+    errorCapture = new ErrorCapture(exceptionUtils);
+    sapAdapterImpl = new SAPAdapterImpl(errorCapture, connection);
+    Map opProps = new HashMap<>();
+    opProps.put("RFC", "ZTEST_DATA_DEL_ACDOCA");
```
what is this?

sumitnigamcs replied:
code unused, so removed

rmstar commented on src/e2e-test/java/io/cdap/plugin/odp/stepsdesign/RuntimeODP.java:
```diff
@@ -0,0 +1,422 @@
+    try {
+      ObjectMapper mapper = new ObjectMapper();
+      ObjectNode objectNode = mapper.createObjectNode();
+      objectNode.put("I_ALL", "X");
```
what is this? please add comments explaining what this means

sumitnigamcs replied:
code unused, so removed

rmstar commented on src/e2e-test/java/io/cdap/plugin/odp/stepsdesign/RuntimeODP.java:
```diff
@@ -0,0 +1,422 @@
+
+  public static void main(String args[]) throws IOException, JCoException, InterruptedException {
+    RuntimeODP runtimeODP = new RuntimeODP();
+    runtimeODP.deleteTheExistingOdpTableInBigquery("tab_src01");
```
please use constants instead of hardcoding table names

sumitnigamcs replied:
done

rmstar commented on src/e2e-test/java/io/cdap/plugin/odp/stepsdesign/RuntimeODP.java:
```diff
@@ -0,0 +1,422 @@
+        Iterator<String> fieldName = object.fieldNames();
+        while (fieldName.hasNext() && zeroIndex == 0) {
+          numList.add(object.get(fieldName.next()).asText());
+          zeroIndex++;
```
if (fieldName.hasNext()) {
  numList.add(object.get(fieldName.next()).asText());
}

sumitnigamcs replied:
since we have multiple elements inside the iterator, thats why while loop

sumitnigamcs replied:
done

rmstar commented on src/e2e-test/java/io/cdap/plugin/odp/stepsdesign/RuntimeODP.java:
```diff
@@ -0,0 +1,422 @@
+      }
+      BeforeActions.scenario.write("No of records :-" + noOfRecords + Arrays.toString(numList.toArray()));
+      Assert.assertEquals(noOfRecords, Integer.parseInt(recordcount));
+     // Thread.sleep(60000);
```
remove commented code

sumitnigamcs replied:
done

rmstar commented on src/e2e-test/java/io/cdap/plugin/odp/stepsdesign/RuntimeODP.java:
```diff
@@ -0,0 +1,422 @@
+      JsonNode response = sapAdapterImpl.executeRFC(objectNode.toString(), opProps,
+                                                    StringUtils.EMPTY, StringUtils.EMPTY);
+    } catch (Exception e) {
+      throw SystemException.throwException(e.getMessage(), e);
```
why you do need to do this?

sumitnigamcs replied:
to abort the scenario in case record is not created

rmstar commented on src/e2e-test/java/io/cdap/plugin/odp/stepsdesign/RuntimeODP.java:
```diff
@@ -0,0 +1,422 @@
+      (table) + "` WHERE " +
+      field.toUpperCase();
+    switch (datasource) {
+      case "0MATERIAL_ATTR":
```
use constants for datasources and actions if possible

sumitnigamcs replied:
done

rmstar commented on src/e2e-test/java/io/cdap/plugin/odp/stepsdesign/RuntimeODP.java:
```diff
@@ -0,0 +1,422 @@
+  static String deltaLog;
+  static String rawLog;
+  static PrintWriter out;
+  private List<String> numList = new ArrayList<>();
```
what is stored in numList? maybe use a more descriptive name?

sumitnigamcs replied:
done

rmstar commented on src/e2e-test/java/io/cdap/plugin/odp/stepsdesign/RuntimeODP.java:
```diff
@@ -0,0 +1,422 @@
+      field.toUpperCase();
+    switch (datasource) {
+      case "0MATERIAL_ATTR":
+        for (i = 0; i < numList.size(); i++) {
```
for (String thing : numList) {
   // do stuff with thing
}

sumitnigamcs replied:
done

rmstar commented on src/e2e-test/java/io/cdap/plugin/odp/stepsdesign/RuntimeODP.java:
```diff
@@ -0,0 +1,422 @@
+        BeforeActions.scenario.write("Table Deleted Successfully");
+      }
+    } catch (Exception e) {
+      BeforeActions.scenario.write(e.toString());
```
should we fail the test if an exception was thrown?

ksahilCF replied:
No we should not fail as if the table is not found in the catch block it will return table not found and we have written the same in the report.

sumitnigamcs committed:
review comments

tivv reviewed (changes requested):
What are the jars added? It's usually an antipattern as we must be very careful on versioning / licensing / availability.

sumitnigamcs committed:
macro changes

sumitnigamcs committed:
license

sumitnigamcs committed:
checkstyle

sumitnigamcs committed:
runner unsequencing

sumitnigamcs committed:
bug fix

sumitnigamcs committed:
secure scenario

sumitnigamcs committed:
rename prop

sumitnigamcs commented:
> What are the jars added? It's usually an antipattern as we must be very careful on versioning / licensing / availability.

Hi @tivv , we have added the jar com.google.adapter.cs-0.0.1.jar, which has JCO connectors for record creation deletion and updation. Please suggest if there is any other way to use the jar through artifactory or something

sumitnigamcs committed:
duplicate plugins

sumitnigamcs committed:
guava moved to test profile

tivv commented:
POMs look good. What's about the jars? I don't think we can include them like this

sumitnigamcs commented:
> POMs look good. What's about the jars? I don't think we can include them like this.
@tivv working on the jars removal now.

sumitnigamcs committed:
jars removed and github action details added to readme

sumitnigamcs commented:
> POMs look good. What's about the jars? I don't think we can include them like this

Removed the jars, and added github actions details to README, and a mention about it in pom. Please have a look
@tivv @rmstar @itsankit-google

sumitnigamcs commented:
@tivv @rmstar @itsankit-google all review comments addressed. Please review, mainly on the jars removal.
